File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.py
#  A = {10, 20, 10, 30, 20, 50} 
# 현재값과 이전 값을 비교해 각 자리마다 cnt를 적용해주자
# 만약 현재값 < 이전값이라면 다시 위로 for문 순회하자
# --> [1,2, ,3, ,4]
# 최종적으로는 max값 출력

import sys
input = sys.stdin.readline
N = int(input())

arr = [0] + list(map(int, input().strip().split()))
dp = [0] * (N+1)


# 1. 이전값 비교 알고리즘
for i in range(1, N+1):
    for j in range(i):
        if arr[i] > arr[j]:
            dp[i] = max(dp[i], dp[j]+1)

# 2. 중간에 최대값이 있을 수도 있기에 max 값 찾기 필수
print(max(dp))

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.BufferedReader;

import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int N = Integer.parseInt(br.readLine());
        int[] arr = new int[N+1];
        int[] dp = new int[N+1];

        st = new StringTokenizer(br.readLine(), " ");
        for (int i = 1; i < N+1; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        // dp
        for (int i = 1; i < N+1; i++) {
            for (int j = 0; j < i; j++) {
                if (arr[i] > arr[j]) {
                    dp[i] = Math.max(dp[i], dp[j]+1);
                }
            }
        }

        // dp 배열에서 최댓값
        int max_sum = 0;
        for (int i = 1; i < N+1; i++) {
            max_sum = Math.max(max_sum, dp[i]);
        }

        System.out.println(max_sum);


    }
}